File: Classes/CustomBadge.m
/*
 
 ------------------------------------------------------------------------------------
 CustomBadge.m
 ------------------------------------------------------------------------------------
 CustomBadge is an UIView which draws a customizable badge on any other view.
 The latest version has separation between style and rendering.
 This class is the core of CustomBadge where the actual rendering happens.
 It recommended to use the convenient allocators instead of the init methods.
 ------------------------------------------------------------------------------------
 
 The MIT License (MIT)
 
 Copyright (c) 2014 Sascha Paulus
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 
 */


#import "CustomBadge.h"


@interface CustomBadge()

@property(nonatomic) UIFont *badgeFont;

- (void) drawRoundedRectWithContext:(CGContextRef)context withRect:(CGRect)rect;
- (void) drawFrameWithContext:(CGContextRef)context withRect:(CGRect)rect;

@end

@implementation CustomBadge

@synthesize badgeText;
@synthesize badgeCornerRoundness;
@synthesize badgeScaleFactor;
@synthesize badgeStyle;

// I recommend to use one of the allocators like customBadgeWithString
- (id) initWithString:(NSString *)badgeString withScale:(CGFloat)scale withStyle:(BadgeStyle*)style
{
	self = [super initWithFrame:CGRectMake(0, 0, 25, 25)];
	if(self!=nil) {
		self.contentScaleFactor = [[UIScreen mainScreen] scale];
		self.backgroundColor = [UIColor clearColor];
		self.badgeText = badgeString;
        self.badgeStyle = style;
		self.badgeCornerRoundness = 0.4;
		self.badgeScaleFactor = scale;
        [self autoBadgeSizeWithString:badgeString];
	}
	return self;
}


// Use this method if you want to change the badge text after the first rendering 
- (void) autoBadgeSizeWithString:(NSString *)badgeString
{
	CGSize retValue;
	CGFloat rectWidth, rectHeight;
    NSDictionary *fontAttr = @{ NSFontAttributeName : [self fontForBadgeWithSize:12] };
	CGSize stringSize = [badgeString sizeWithAttributes:fontAttr];
	CGFloat flexSpace;
	if ([badgeString length]>=2) {
		flexSpace = [badgeString length];
		rectWidth = 25 + (stringSize.width + flexSpace); rectHeight = 25;
		retValue = CGSizeMake(rectWidth*badgeScaleFactor, rectHeight*badgeScaleFactor);
	} else {
		retValue = CGSizeMake(25*badgeScaleFactor, 25*badgeScaleFactor);
	}
	self.frame = CGRectMake(self.frame.origin.x, self.frame.origin.y, retValue.width, retValue.height);
	self.badgeText = badgeString;
	[self setNeedsDisplay];
}


// Creates a Badge with a given Text in default BadgeStyle and normal scale
+ (CustomBadge*) customBadgeWithString:(NSString *)badgeString
{
    return [[self alloc] initWithString:badgeString withScale:1.0 withStyle:[BadgeStyle defaultStyle]];
}

// Creates a Badge with a given Text in default BadgeStyle and given scale
+ (CustomBadge*) customBadgeWithString:(NSString *)badgeString withScale:(CGFloat)scale {
    
    return [[self alloc] initWithString:badgeString withScale:scale withStyle:[BadgeStyle defaultStyle]];
    
}

// Creates a Badge with a given Text in given BadgeStyle and normal scale
+ (CustomBadge*) customBadgeWithString:(NSString *)badgeString withStyle:(BadgeStyle*)style
{
    return [[self alloc] initWithString:badgeString withScale:1.0 withStyle:style];
}


// Creates a Badge with a given Text in given BadgeStyle and a given scale
+ (CustomBadge*) customBadgeWithString:(NSString *)badgeString withScale:(CGFloat)scale withStyle:(BadgeStyle*)style {

    return [[self alloc] initWithString:badgeString withScale:scale withStyle:style];
    
}
 

// Draws the Badge with Quartz
-(void) drawRoundedRectWithContext:(CGContextRef)context withRect:(CGRect)rect
{
	CGContextSaveGState(context);
	
	CGFloat radius = CGRectGetMaxY(rect)*self.badgeCornerRoundness;
	CGFloat puffer = CGRectGetMaxY(rect)*0.10;
	CGFloat maxX = CGRectGetMaxX(rect) - puffer;
	CGFloat maxY = CGRectGetMaxY(rect) - puffer;
	CGFloat minX = CGRectGetMinX(rect) + puffer;
	CGFloat minY = CGRectGetMinY(rect) + puffer;
		
    CGContextBeginPath(context);
	CGContextSetFillColorWithColor(context, [self.badgeStyle.badgeInsetColor CGColor]);
	CGContextAddArc(context, maxX-radius, minY+radius, radius, M_PI+(M_PI/2), 0, 0);
	CGContextAddArc(context, maxX-radius, maxY-radius, radius, 0, M_PI/2, 0);
	CGContextAddArc(context, minX+radius, maxY-radius, radius, M_PI/2, M_PI, 0);
	CGContextAddArc(context, minX+radius, minY+radius, radius, M_PI, M_PI+M_PI/2, 0);
    if (self.badgeStyle.badgeShadow) {
        CGContextSetShadowWithColor(context, CGSizeMake(1.0,1.0), 3, [[UIColor blackColor] CGColor]);
    }
	CGContextFillPath(context);

	CGContextRestoreGState(context);

}

// Draws the Badge Shine with Quartz
-(void) drawShineWithContext:(CGContextRef)context withRect:(CGRect)rect
{
	CGContextSaveGState(context);
 
	CGFloat radius = CGRectGetMaxY(rect)*self.badgeCornerRoundness;
	CGFloat puffer = CGRectGetMaxY(rect)*0.10;
	CGFloat maxX = CGRectGetMaxX(rect) - puffer;
	CGFloat maxY = CGRectGetMaxY(rect) - puffer;
	CGFloat minX = CGRectGetMinX(rect) + puffer;
	CGFloat minY = CGRectGetMinY(rect) + puffer;
	CGContextBeginPath(context);
	CGContextAddArc(context, maxX-radius, minY+radius, radius, M_PI+(M_PI/2), 0, 0);
	CGContextAddArc(context, maxX-radius, maxY-radius, radius, 0, M_PI/2, 0);
	CGContextAddArc(context, minX+radius, maxY-radius, radius, M_PI/2, M_PI, 0);
	CGContextAddArc(context, minX+radius, minY+radius, radius, M_PI, M_PI+M_PI/2, 0);
	CGContextClip(context);
	
	
	size_t num_locations = 2;
	CGFloat locations[2] = { 0.0, 0.4 };
	CGFloat components[8] = {  0.92, 0.92, 0.92, 1.0, 0.82, 0.82, 0.82, 0.4 };

	CGColorSpaceRef cspace;
	CGGradientRef gradient;
	cspace = CGColorSpaceCreateDeviceRGB();
	gradient = CGGradientCreateWithColorComponents (cspace, components, locations, num_locations);
	
	CGPoint sPoint, ePoint;
	sPoint.x = 0;
	sPoint.y = 0;
	ePoint.x = 0;
	ePoint.y = maxY;
	CGContextDrawLinearGradient (context, gradient, sPoint, ePoint, 0);
	
	CGColorSpaceRelease(cspace);
	CGGradientRelease(gradient);
	
	CGContextRestoreGState(context);	
}


// Draws the Badge Frame with Quartz
-(void) drawFrameWithContext:(CGContextRef)context withRect:(CGRect)rect
{
	CGFloat radius = CGRectGetMaxY(rect)*self.badgeCornerRoundness;
	CGFloat puffer = CGRectGetMaxY(rect)*0.10;
	
	CGFloat maxX = CGRectGetMaxX(rect) - puffer;
	CGFloat maxY = CGRectGetMaxY(rect) - puffer;
	CGFloat minX = CGRectGetMinX(rect) + puffer;
	CGFloat minY = CGRectGetMinY(rect) + puffer;
	
	
    CGContextBeginPath(context);
	CGFloat lineSize = 2;
	if(self.badgeScaleFactor>1) {
		lineSize += self.badgeScaleFactor*0.25;
	}
	CGContextSetLineWidth(context, lineSize);
	CGContextSetStrokeColorWithColor(context, [self.badgeStyle.badgeFrameColor CGColor]);
	CGContextAddArc(context, maxX-radius, minY+radius, radius, M_PI+(M_PI/2), 0, 0);
	CGContextAddArc(context, maxX-radius, maxY-radius, radius, 0, M_PI/2, 0);
	CGContextAddArc(context, minX+radius, maxY-radius, radius, M_PI/2, M_PI, 0);
	CGContextAddArc(context, minX+radius, minY+radius, radius, M_PI, M_PI+M_PI/2, 0);
	CGContextClosePath(context);
	CGContextStrokePath(context);
}


- (UIFont*) fontForBadgeWithSize:(CGFloat)size {
    switch (self.badgeStyle.badgeFontType) {
        case BadgeStyleFontTypeHelveticaNeueMedium:
            return [UIFont fontWithName:@"HelveticaNeue-Medium" size:size];
            break;
        default:
            return [UIFont fontWithName:@"HelveticaNeue-Light" size:size];
            break;
    }
}

- (void)drawRect:(CGRect)rect {
	
	CGContextRef context = UIGraphicsGetCurrentContext();
	[self drawRoundedRectWithContext:context withRect:rect];
	
	if(self.badgeStyle.badgeShining) {
		[self drawShineWithContext:context withRect:rect];
	}
	
	if (self.badgeStyle.badgeFrame)  {
		[self drawFrameWithContext:context withRect:rect];
	}
	
	if ([self.badgeText length]>0) {
		CGFloat sizeOfFont = 13.5*badgeScaleFactor;
		if ([self.badgeText length]<2) {
            sizeOfFont += sizeOfFont * 0.20f;
		}
        UIFont *textFont =  [self fontForBadgeWithSize:sizeOfFont];
        NSDictionary *fontAttr = @{ NSFontAttributeName : textFont, NSForegroundColorAttributeName : self.badgeStyle.badgeTextColor };
		CGSize textSize = [self.badgeText sizeWithAttributes:fontAttr];
        CGPoint textPoint = CGPointMake((rect.size.width/2-textSize.width/2), (rect.size.height/2-textSize.height/2) - 1 );
		[self.badgeText drawAtPoint:textPoint withAttributes:fontAttr];
	}
}




@end

File: Classes/Swift/CustomBadge.swift
/*
CustomBadge.m

*** Description: ***
With this class you can draw a typical iOS badge indicator with a custom text on any view.
Please use the allocator customBadgeWithString to create a new badge.
In this version you can modfiy the color inside the badge (insetColor),
the color of the frame (frameColor), the color of the text and you can
tell the class if you want a frame around the badge.

*** License & Copyright ***
Created by Sascha Paulus www.spaulus.com on 04/2011. Version 2.0
This tiny class can be used for free in private and commercial applications.
Please feel free to modify, extend or distribution this class.
If you modify it: Please send me your modified version of the class.
Please do not sell the source code solely and keep this text in
your copyright section. Thanks.

If you have any questions please feel free to contact me (open@spaulus.com).

*/

import UIKit

class CustomBadge: UIControl {
    
    var badgeText : String  = ""
    var badgeTextColor : UIColor = UIColor.blackColor()
    var badgeInsetColor : UIColor = UIColor.blackColor()
    var badgeFrameColor : UIColor = UIColor.blackColor()
    var badgeFrame : Bool = false;
    var badgeCornerRoundness : CGFloat = 1.0
    var badgeScaleFactor : CGFloat = 1.0
    var badgeShining : Bool = false
    
    var objectTag : AnyObject?
    
    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // I recommend to use the allocator customBadgeWithString
    init(badgeString: String, frame: CGRect, scale: Float, shining: Bool) {
        super.init(frame: frame)
    
        setUp(badgeString: badgeString, stringColor: UIColor.whiteColor(), insetColor: UIColor.redColor(), hasBadgeFrame: true,
            frameColor: UIColor.whiteColor(), scale: scale, shining: shining)
    }
    
    init(badgeString: String, frame: CGRect, stringColor: UIColor, insetColor: UIColor, hasBadgeFrame: Bool, frameColor: UIColor,
        scale: Float, shining: Bool) {
        super.init(frame: frame)
        
        setUp(badgeString: badgeString, stringColor: stringColor, insetColor: insetColor, hasBadgeFrame: hasBadgeFrame, frameColor: frameColor, scale: scale, shining: shining)
    }
    
    func setUp(#badgeString: String, stringColor: UIColor, insetColor: UIColor, hasBadgeFrame: Bool, frameColor: UIColor,
        scale: Float, shining: Bool) {
        
        self.contentScaleFactor = UIScreen.mainScreen().scale
        self.backgroundColor = UIColor.clearColor()
        self.badgeText = badgeString
        self.badgeTextColor = stringColor
        self.badgeFrame = hasBadgeFrame;
        self.badgeFrameColor = frameColor
        self.badgeInsetColor = insetColor
        self.badgeCornerRoundness = CGFloat(0.4)
        self.badgeScaleFactor = CGFloat(scale)
        self.badgeShining = shining
        self.autoBadgeSize(badgeString)
    }
    
    // Use this method if you want to change the badge text after the first rendering
    func autoBadgeSize(badgeString: String) {
        var retValue : CGSize
        var rectWidth, rectHeight : CGFloat
        var stringSize = (badgeString as NSString).sizeWithAttributes([NSFontAttributeName: UIFont.boldSystemFontOfSize(12)])
        var flexSpace : CGFloat
        
        let badgeStringLength = (badgeString as NSString).length
        
        if (badgeStringLength >= 2) {
            flexSpace = CGFloat(badgeStringLength)
            rectWidth = 25 + (stringSize.width + flexSpace);
            rectHeight = CGFloat(25);
            
            retValue = CGSizeMake(CGFloat(rectWidth * badgeScaleFactor), CGFloat(rectHeight * badgeScaleFactor));
        } else {
            retValue = CGSizeMake(25 * badgeScaleFactor, 25 * badgeScaleFactor);
        }
        
        self.frame = CGRectMake(self.frame.origin.x, self.frame.origin.y, retValue.width, retValue.height)
        self.badgeText = badgeString
        self.setNeedsDisplay()
    }
    
    
    // Creates a Badge with a given Text
    class func customBadge(badgeString: String, frame: CGRect) -> CustomBadge {
        return CustomBadge(badgeString: badgeString, frame: frame, scale: 1.0, shining: true)
    }
    
    
    // Creates a Badge with a given Text, Text Color, Inset Color, Frame (YES/NO) and Frame Color
    class func customBadge(badgeString: String, frame: CGRect, stringColor: UIColor, insetColor: UIColor, hasBadgeFrame: Bool,
        badgeFrameColor: UIColor, scale: Float, shining: Bool) -> CustomBadge {
            
            return CustomBadge(badgeString: badgeString, frame: frame, stringColor: stringColor, insetColor: insetColor,
                hasBadgeFrame: hasBadgeFrame, frameColor: badgeFrameColor, scale: scale, shining: shining)
    }
    
    
    
    
    
    // Draws the Badge with Quartz
    func drawRoundedRect(context: CGContextRef, rect: CGRect) {
        CGContextSaveGState(context)
    
        let radius = CGRectGetMaxY(rect) * badgeCornerRoundness
        let puffer = CGRectGetMaxY(rect) * 0.10
        let maxX = CGRectGetMaxX(rect) - puffer
        let maxY = CGRectGetMaxY(rect) - puffer
        let minX = CGRectGetMinX(rect) + puffer
        let minY = CGRectGetMinY(rect) + puffer
    
        CGContextBeginPath(context)
        CGContextSetFillColorWithColor(context, badgeInsetColor.CGColor)
        
        CGContextAddArc(context, maxX - radius, minY + radius,
            radius, CGFloat(M_PI + (M_PI / 2)), 0, 0)
        
        CGContextAddArc(context, maxX - radius, maxY - radius, radius, 0, CGFloat(M_PI / 2), 0)
        
        CGContextAddArc(context, minX + radius, maxY - radius, radius, CGFloat(M_PI / 2), CGFloat(M_PI), 0)
        
        CGContextAddArc(context, minX + radius, minY + radius, radius, CGFloat(M_PI), CGFloat(M_PI + M_PI / 2), 0)
        
        CGContextSetShadowWithColor(context, CGSizeMake(1.0, 1.0), 3, UIColor.blackColor().CGColor)
        CGContextFillPath(context)
    
        CGContextRestoreGState(context)
    }
    
    // Draws the Badge Shine with Quartz
    func drawShine(context: CGContextRef, rect: CGRect) {
    
        CGContextSaveGState(context)
    
        let radius = CGRectGetMaxY(rect) * badgeCornerRoundness
        let puffer = CGRectGetMaxY(rect) * 0.10
        let maxX = CGRectGetMaxX(rect) - puffer
        let maxY = CGRectGetMaxY(rect) - puffer
        let minX = CGRectGetMinX(rect) + puffer
        let minY = CGRectGetMinY(rect) + puffer
        
        CGContextBeginPath(context)
        
        CGContextAddArc(context, maxX - radius, minY + radius, radius, CGFloat(M_PI + (M_PI / 2)), 0, 0)
        
        CGContextAddArc(context, maxX - radius, maxY - radius, radius, 0, CGFloat(M_PI / 2), 0)
        
        CGContextAddArc(context, minX + radius, maxY - radius, radius, CGFloat(M_PI / 2), CGFloat(M_PI), 0)
        
        CGContextAddArc(context, minX + radius, minY + radius, radius, CGFloat(M_PI), CGFloat(M_PI + M_PI / 2), 0)
        
        CGContextClip(context)
    
        let num_locations = size_t(2)
        let locations : [CGFloat] = [ 0.0, 0.4 ]
        let components : [CGFloat] = [0.92, 0.92, 0.92, 1.0, 0.82, 0.82, 0.82, 0.4]
    
        let cspace = CGColorSpaceCreateDeviceRGB()
        let gradient = CGGradientCreateWithColorComponents(cspace, components, locations, num_locations)
    
        let sPoint = CGPoint(x: 0, y: 0)
        let ePoint = CGPoint(x: 0, y: maxY)
        
        CGContextDrawLinearGradient (context, gradient, sPoint, ePoint, 0)
    
        CGContextRestoreGState(context)
    }
    
    
    // Draws the Badge Frame with Quartz
    func drawFrame(context: CGContextRef, rect: CGRect) {
        
        let radius = CGRectGetMaxY(rect) * badgeCornerRoundness
        let puffer = CGRectGetMaxY(rect) * 0.10
    
        let maxX = CGRectGetMaxX(rect) - puffer
        let maxY = CGRectGetMaxY(rect) - puffer
        let minX = CGRectGetMinX(rect) + puffer
        let minY = CGRectGetMinY(rect) + puffer
    
    
        CGContextBeginPath(context)
        var lineSize = CGFloat(2)
        
        if (badgeScaleFactor > 1) {
            lineSize += badgeScaleFactor * 0.25
        }
        
        CGContextSetLineWidth(context, lineSize)
        CGContextSetStrokeColorWithColor(context, badgeFrameColor.CGColor)
        CGContextAddArc(context, maxX - radius, minY + radius, radius, CGFloat(M_PI + (M_PI / 2)), 0, 0)
        CGContextAddArc(context, maxX - radius, maxY - radius, radius, 0, CGFloat(M_PI / 2), 0)
        CGContextAddArc(context, minX + radius, maxY - radius, radius, CGFloat(M_PI / 2), CGFloat(M_PI), 0)
        CGContextAddArc(context, minX + radius, minY + radius, radius, CGFloat(M_PI), CGFloat(M_PI + M_PI / 2), 0)
        
        CGContextClosePath(context)
        CGContextStrokePath(context)
    }
    
    
    override func drawRect(rect: CGRect) {
    
        let context = UIGraphicsGetCurrentContext()
        drawRoundedRect(context, rect: rect)
    
        if (badgeShining) {
            drawShine(context, rect:rect)
        }
    
        if (badgeFrame)  {
            drawFrame(context, rect:rect)
        }
    
        if ((badgeText as NSString).length > 0) {
            
            badgeTextColor.set()
            
            var sizeOfFont = 13.5 * badgeScaleFactor
            
            if ((badgeText as NSString).length < 2) {
                sizeOfFont += sizeOfFont * 0.20;
            }
            
            let textFont = UIFont.boldSystemFontOfSize(sizeOfFont)
            
            let textSize = (badgeText as NSString).sizeWithAttributes([NSFontAttributeName: textFont])
            
            (badgeText as NSString).drawAtPoint(CGPointMake((rect.size.width/2-textSize.width/2), (rect.size.height/2-textSize.height/2)), withAttributes: [NSFontAttributeName: textFont])
        }
    
    }

}
